File: .claude/skills/zero-trust-secret-orchestration/references/mjs/src/strategies/keychain-biometric.mjs
import { Strategy, SecretResolutionError } from "../strategy.mjs";
import { runSandboxed } from "../harness.mjs";

/**
 * Exit code mapping for `security find-generic-password -w`:
 *   0   — success; stdout is the password (trim trailing newline).
 *   44  — keychain item not found → resolver should continue chain → null.
 *   128 — user cancelled the biometric prompt → SecretResolutionError.
 *   other — unexpected; treat as resolution failure → SecretResolutionError.
 */
export class KeychainBiometricStrategy extends Strategy {
  constructor({ service, _runner } = {}) {
    super();
    this.service = service ?? process.env.KEYCHAIN_SERVICE ?? "zts";
    this._runner = _runner ?? runSandboxed;
  }

  async get(key) {
    const { stdout, code } = await this._runner(
      "security",
      ["find-generic-password", "-s", this.service, "-a", key, "-w"],
      { timeoutMs: 30_000 },
    );
    if (code === 44) return null;
    if (code === 128) {
      throw new SecretResolutionError("biometric prompt cancelled by user");
    }
    if (code !== 0) {
      throw new SecretResolutionError(`security exited ${code}`);
    }
    return stdout.toString().trim();
  }
}

File: .claude/skills/zero-trust-secret-orchestration/references/py/src/zts/strategies/keychain_biometric.py
import os

from zts.harness import run_sandboxed
from zts.strategy import SecretResolutionError, Strategy


class KeychainBiometricStrategy(Strategy):
    """macOS Keychain reader gated by Touch ID via `security`.

    Exit code mapping for `security find-generic-password -w`:
      0   — success; stdout is the password.
      44  — keychain item not found → None.
      128 — user cancelled the biometric prompt → SecretResolutionError.
      other — unexpected → SecretResolutionError.
    """

    def __init__(self, service: str | None = None, *, _runner=None):
        self.service = service or os.environ.get("KEYCHAIN_SERVICE") or "zts"
        self._runner = _runner or run_sandboxed

    async def get(self, key: str) -> str | None:
        result = await self._runner(
            "security",
            "find-generic-password",
            "-s",
            self.service,
            "-a",
            key,
            "-w",
            timeout_ms=30_000,
        )
        if result.code == 44:
            return None
        if result.code == 128:
            raise SecretResolutionError("biometric prompt cancelled by user")
        if result.code != 0:
            raise SecretResolutionError(f"security exited {result.code}")
        return result.stdout.decode().rstrip("\n")
